Make leaderboard tabs and active ruleset filter scores

naoey opened the pull request:
One thing I'm not certain of is the conditions for the `mode` parameter of the request. I've noticed trying to fetch with `mode=osu` for a taiko beatmap returns 403 and the api disconnects. I didn't get around to discerning what scores stable fetches for a taiko beatmap in osu mode, but shouldn't the carousel show only beatmaps playable in the current ruleset?

naoey committed:
Add respective query params to GetScoreRequest based on selected tab.

naoey committed:
Add placeholder when there are no scores.

naoey committed:
Move loader hiding to a better place.

naoey committed:
Add game mode query to request.

- Also update scores when game mode is changed

naoey committed:
Handle query params better.

naoey committed:
Add a retry button for when scores request fails.

naoey committed:
Merge remote-tracking branch 'upstream/master' into leaderboard-scopes

naoey committed:
Don't break VisualTests and add a real beatmap step.

naoey committed:
Unbind ruleset event from leaderboard.

naoey committed:
Remove unnecessary null check and tweak transform a bit.

naoey committed:
Put retry button in a BeatSyncedContainer and change error message.

naoey committed:
Use a single placeholder container for empty and retry.

naoey committed:
R#

peppy commented on osu.Game/Online/API/Requests/GetScoresRequest.cs:
```diff
@@ -27,12 +31,67 @@ public GetScoresRequest(BeatmapInfo beatmap)
+                    break;
+            }
+
+            switch (ruleset?.Name)
```
we never want these constants to exist in the code. either the api should be changed to accept the ID rather than a name, or we should store these names as an `OnlineIdentifier` or something inside `RulesetInfo` instead.

naoey replied:
yeah it felt wrong but was the only way that i could see to get it working immediately. which would be the preferred solution?

peppy replied:
The API should be changed to support IDs.

peppy replied:
Tracking at https://github.com/ppy/osu-web/issues/1819

peppy commented on osu.Game.Tests/Visual/TestCaseLeaderboard.cs:
```diff
@@ -6,15 +6,43 @@
+            AddStep(@"New Scores", newScores);
+            AddStep(@"Empty Scores", () => leaderboard.Scores = Enumerable.Empty<Score>());
+            AddStep(@"Real beatmap", realBeatmap);
+            newScores();
```
pointless; first step is always run on load

peppy commented on osu.Game.Tests/Visual/TestCaseLeaderboard.cs:
```diff
@@ -6,15 +6,43 @@
 namespace osu.Game.Tests.Visual
 {
     [Description("PlaySongSelect leaderboard")]
     internal class TestCaseLeaderboard : OsuTestCase
```
testcase doesn't test the fail scenario

naoey committed:
Fix line endings and derp that was causing request failures.

naoey committed:
Add failure test case.

- Only show failure if request wasn't cancelled

naoey committed:
Make retry button not look drunk.

naoey commented:
So if I understand correctly, I now need to add an `OnlineIdentifier` property in `RulesetInfo`, add the identifier in each `Ruleset`, and add a migration which adds the `OnlineIdentifier` column?

peppy commented:
That does need to be done, in a separate PR.

peppy commented:
Should be updated after #1684 is merged.

peppy commented:
`RulesetInfo`s now have a `ShortName`. wanna update this to work with that?

naoey committed:
Merge remote-tracking branch 'upstream/master' into leaderboard-scopes

naoey committed:
Use ShortName for mode.

- Also set the ruleset when fetching scores in BeatmapSetOverlay

peppy committed:
Merge branch 'master' into leaderboard-scopes

peppy committed:
Throw an exception if we try and retrieve local scores online

peppy committed:
Simplify scope logic

peppy committed:
Remove unnecessary constructure and make ruleset required

peppy committed:
Fix test edge cases

peppy committed:
Remove unnecessary workaround

peppy commented on osu.Game/Screens/Select/Leaderboards/Leaderboard.cs:
```diff
@@ -38,15 +50,38 @@ public IEnumerable<Score> Scores
 
+                if (!scores.Any())
+                {
+                    placeholderFlow.Children = new Drawable[]
```
instead of setting children here, just make:

```
private class NoScoresPlaceholder : Container (or FlowContainer or whatever)
{
    ctor { Children = { SpriteIcon, OsuSpriteText } }
}

private class RetrievalFailPlaceholder : Container (or FlowContainer or whatever)
{
    ctor { Children = { RetryIcon, OsuSpriteText } }
}
```

and change this code to

```
placeholderContainer.Clear(true);
placeholderContainer.Child = new NoScoresPlaceholder();
placeholderContainer.FadeInFromZero(500, EasingTypes.OutQuint);
```
or so

to tidy things up

peppy commented on osu.Game/Screens/Select/Leaderboards/Leaderboard.cs:
```diff
@@ -83,14 +131,31 @@ public Leaderboard()
+                        placeholderFlow = new FillFlowContainer
+                        {
+                            Direction = FillDirection.Horizontal,
+                            Anchor = Anchor.Centre,
```
Move anchor/origin to placeholderContainer and make it AutoSize, not relative

peppy commented on osu.Game/Screens/Select/Leaderboards/Leaderboard.cs:
```diff
@@ -164,5 +276,85 @@ protected override void Update()
             }
         }
+
+        private class RetryButton : BeatSyncedContainer
```
please remove this bouncing for now. the scale on hover is fine but the bouncing is ugly.

you can probably make this container an `: OsuClickableContainer` itself as a result

peppy commented on osu.Game/Screens/Select/Leaderboards/Leaderboard.cs:
```diff
@@ -164,5 +276,85 @@ protected override void Update()
+                this.colours = colours;
+            }
+
+            protected override void OnNewBeat(int beatIndex, TimingControlPoint timingPoint, EffectControlPoint effectPoint, TrackAmplitudes amplitudes)
```
👎

peppy commented on osu.Game/Screens/Select/Leaderboards/Leaderboard.cs:
```diff
@@ -164,5 +276,85 @@ protected override void Update()
+
+            protected override void OnHoverLost(InputState state)
+            {
+                icon.ClearTransforms();
```
shouldn't be necessary

peppy commented on osu.Game/Screens/Select/Leaderboards/Leaderboard.cs:
```diff
@@ -164,5 +276,85 @@ protected override void Update()
+
+            protected override bool OnMouseDown(InputState state, MouseDownEventArgs args)
+            {
+                icon.FadeColour(colours.Yellow, 400);
```
if anything i'd reverse the face and hover transforms in line with the rest of lazer..

in fact, if you make this whole class `: OsuHoverContainer` this will perform the on-hover fade for you.

then cahnge mousedown to scaleTo(0.8) and mouseup to scaleTo(1.0) or something similar to OsuButton

naoey committed:
Create placeholder classes instead of changing Children.

- Add MessagePlaceholder
- Use MessagePlacholder for when API is offline/user isn't a supporter
- Remove unnecessary placeholderFlow field
- Hook into API state changes

naoey committed:
Kill ugly retry button bounce.

naoey committed:
Remove unused usings.

naoey committed:
Increase GetScoresRequest timeout & leave existing scores when API dies.

peppy committed:
Show retrieval failure when OnlineBeatmapID is missing

peppy committed:
Reorder API / logged in checks to make more sense

naoey committed:
Maintain a placeholder state and add tests showing all the states.

- Also don't replace placeholder if new one is same as old

peppy committed:
Merge branch 'master' into leaderboard-scopes

naoey committed:
Unsubscribe from API state change event.

peppy committed:
Fix minor formatting issue

peppy commented on osu.Game/Screens/Select/Leaderboards/Leaderboard.cs:
```diff
@@ -109,33 +179,120 @@ public BeatmapInfo Beatmap
+            this.osuGame = osuGame;
+
+            if (osuGame != null)
+                osuGame.Ruleset.ValueChanged += handleRulesetChange;
```
For bindables, rather than binding/unbinding to global bindables, you should create a local bindable `private readonly Bindable<Ruleset> ruleset = new Bindable<Ruleset>(); ` and use `ruleset.BindTo(osuGame.Ruleset)`

peppy replied:
you don't need to unbind `ValueChanged` if you do this.

for the api unfortunately you will have to do the unbind because `OnStateChange` is not a bindable currently.

peppy committed:
Use local bindable

peppy committed:
Remove unnecessary IsLoaded check

We are always loaded at this point.

peppy committed:
Merge branch 'master' into leaderboard-scopes

peppy committed:
Adjust debounce a bit

peppy committed:
Adjust animation slightly

peppy commented:
Good job on this one. Finished up the last piece myself so we can merge this in 👍 .
